src/Psl/Regex/replace.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Regex;
+
+use Psl\Internal;
+use Psl\Str;
+use Psl\Type;
+
+use function preg_replace;
+
+/**
+ * Returns the '$haystack' string with all occurrences of `$pattern` replaced by
+ * `$replacement`.
+ *
+ * @param non-empty-string  $pattern    The pattern to search for.
+ *
+ * @throws Exception\InvalidPatternException If $pattern is invalid.
+ * @throws Exception\RuntimeException In case of an unexpected error.
+ *
+ * @psalm-pure
+ */
+function replace(string $haystack, string $pattern, string $replacement): string
+{
+    $result = Internal\suppress(
+        static fn() => preg_replace($pattern, $replacement, $haystack)
+    );
+
+    if ($result->isFailed()) {
```

Comment:
The code looks good, but just wondering:

WDYT about using the power of the result classes ?
It would safe you some psalm suppresses and might make the flow more visual. 
Now it's a result, then an if with error handling and finally a coercion within a try.

IMO, this code makes the flow a bit more understandable.

```php
function replace(string $haystack, string $pattern, string $replacement): string
{
    return Internal\suppress(
        static fn() => preg_replace($pattern, $replacement, $haystack)
    )->then(
        static fn ($result): string => Type\string()->assert($result),
        static function (Exception $previous) {
            throw new Exception\InvalidPatternException(Str\replace(
                $previous->getMessage(),
                'preg_replace(): ',
                ''
            ), 0, $previous);
        } 
    )->proceed(
        Fun\identity(),
        static function (Exception $error) {
            if ($error instanceof  Type\Exception\AssertException) {
                throw new Exception\RuntimeException('Unexpected error', 0, $e);
            }
            
            throw $error;
        }
    );
}

````

Reply:
will not save us from suppression ( `then` and `proceed` are not pure ) :/

Reply:
simplified the code here a bit.